processing/src/main/java/io/druid/query/groupby/GroupByQuery.java

```diff
@@ -300,10 +323,172 @@ public int compare(Object lhs, Object rhs)
     );
   }
 
-  public Ordering<Row> getRowOrdering(final boolean granular)
+  private boolean validateAndGetForceLimitPushDown()
+  {
+    final boolean forcePushDown = getContextBoolean(GroupByQueryConfig.CTX_KEY_FORCE_LIMIT_PUSH_DOWN, false);
+    if (forcePushDown) {
+      if (!(limitSpec instanceof DefaultLimitSpec)) {
+        throw new IAE("When forcing limit push down, a limit spec must be provided.");
+      }
+
+      if (((DefaultLimitSpec) limitSpec).getLimit() == Integer.MAX_VALUE) {
+        throw new IAE("When forcing limit push down, the provided limit spec must have a limit.");
+      }
+
+      for (OrderByColumnSpec orderBySpec : ((DefaultLimitSpec) limitSpec).getColumns()) {
+        if (OrderByColumnSpec.getPostAggIndexForOrderBy(orderBySpec, postAggregatorSpecs) > -1) {
+          throw new UnsupportedOperationException("Limit push down when sorting by a post aggregator is not supported.");
+        }
+      }
+    }
+    return forcePushDown;
+  }
+
+  public boolean determineApplyLimitPushDown()
```

Comment:
Hm, I decided to change this area in a different way, I removed an unnecessary null and type check from sortingOrderHasNonGroupingFields(), but kept determineApplyLimitPushDown() the same.

The sortingOrder method is used elsewhere, where the MAX_VALUE check is no longer relevant (as it would've been already checked earlier when the query object was created), so I didn't think it really belongs there.

 I also felt that the conditions for when limit push down will be applied are more clear if they're all in the determineApplyLimitPushDown() method.